### PR TITLE
fix: resolve issue with conversion string to simple-json

### LINKED ISSUE #4440: "SyntaxError: Unexpected token o in JSON at position 1" when using `@Column("simple-json")`
**Issue type:**

[ ] question
[x] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[x] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[ ] `latest`
[ ] `@next`
[x] `"typeorm": "^0.2.17"`

**Steps to reproduce or a small repository showing the problem:**

Well, I'm trying to make an array for a 'custom field' feature of my backend. In my model, I have
```typescript
  @Column("simple-json", { nullable: true })
  public customFields!: { name: string, content: string };
```
which seems correct, given the documentation.

However, then running...
```
char.customFields = req.body.customFields ? req.body.customFields : char.customFields;
```
results in this error.

```
SyntaxError: Unexpected token o in JSON at position 1
    at JSON.parse (<anonymous>)
    at Function.DateUtils.stringToSimpleJson (C:\Users\Hayden\Source\Repos\roleplay-system\api\src\util\DateUtils.ts:173:49)
    at PostgresDriver.prepareHydratedValue (C:\Users\Hayden\Source\Repos\roleplay-system\api\src\driver\postgres\PostgresDriver.ts:483:31)
    at C:\Users\Hayden\Source\Repos\roleplay-system\api\src\query-builder\transformer\RawSqlResultsToEntityTransformer.ts:143:55
    at Array.forEach (<anonymous>)
    at RawSqlResultsToEntityTransformer.transformColumns (C:\Users\Hayden\Source\Repos\roleplay-system\api\src\query-builder\transformer\RawSqlResultsToEntityTransformer.ts:128:26)
    at RawSqlResultsToEntityTransformer.transformRawResultsGroup (C:\Users\Hayden\Source\Repos\roleplay-system\api\src\query-builder\transformer\RawSqlResultsToEntityTransformer.ts:105:33)
    at C:\Users\Hayden\Source\Repos\roleplay-system\api\src\query-builder\transformer\RawSqlResultsToEntityTransformer.ts:43:33
    at Map.forEach (<anonymous>)
    at RawSqlResultsToEntityTransformer.transform (C:\Users\Hayden\Source\Repos\roleplay-system\api\src\query-builder\transformer\RawSqlResultsToEntityTransformer.ts:42:15)
```

This seems to be something that happened before, back in November 2016. I read #96 and it seems like it was fixed. However... it now seems like the objects hitting TypeORM are still objects, so they're either not being stringified (which I did try doing manually to no success) or something else is going on here.

If it's my code error, *please* let me know.

### PR DESCRIPTION
**Pull request by abhijeet1403**
Closes  #4440

### CODE REVIEW

**Comment by vlapo** [comment]
Please update code style.

**Comment by abhijeet1403** [comment]
Sure.

**Comment by abhijeet1403** [comment]
@vlapo Can you merge it now?

**Comment by abhijeet1403** [comment]
@vlapo  If i think about it, this method returns the raw value or a JSON object, don't you think it should always return a JSON object. Regardless.

**Comment by vlapo** [comment]
Yeah I think we should return empty json object if parse string throw error.

File: test/github-issues/4440/issue-4440.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+describe("github issues > #4440 simple-json column type throws error for string with no value", () => {
+
+    let connections: Connection[];
+    before(async () => {
+        connections = await createTestingConnections({
+            entities: [__dirname + "/entity/*{.js,.ts}"],
+            schemaCreate: true,
+            dropSchema: true,
+            enabledDrivers: ["sqlite", "mysql"],
```
[the remark this answers is not in the document]
Comment (abhijeet1403) [reply test/github-issues/4440/issue-4440.ts]: Added

File: src/util/DateUtils.ts [hunk 2]
```diff
@@ -170,8 +170,10 @@ export class DateUtils {
     }
 
     static stringToSimpleJson(value: any) {
-        // Checking if value is truthy then checking if it is a string
-        return !!value && typeof value === "string" ? JSON.parse(value) : {};
+        // try/catch implemented
+        let simpleJSON = null;
+        try {simpleJSON = JSON.parse(value); } 
+        finally {return (typeof simpleJSON === "object") ? simpleJSON : {}; }
```
Comment (rustamwin) [inline src/util/DateUtils.ts]: Please format your code

File: src/util/DateUtils.ts [hunk 3]
```diff
@@ -170,7 +170,10 @@ export class DateUtils {
     }
 
     static stringToSimpleJson(value: any) {
-        return typeof value === "string" ? JSON.parse(value) : value;
+        // try/catch implemented
+        let simpleJSON = null;
+        try {simpleJSON = JSON.parse(value); }
```
Comment (rustamwin) [inline src/util/DateUtils.ts]:
```suggestion
        try {
            simpleJSON = JSON.parse(value); 
        } 
```

File: src/util/DateUtils.ts [hunk 4]
```diff
@@ -170,7 +170,10 @@ export class DateUtils {
     }
 
     static stringToSimpleJson(value: any) {
-        return typeof value === "string" ? JSON.parse(value) : value;
+        // try/catch implemented
+        let simpleJSON = null;
+        try {simpleJSON = JSON.parse(value); } 
+        finally {return (simpleJSON && typeof simpleJSON === "object") ? simpleJSON : {}; }
```
Comment (rustamwin) [inline src/util/DateUtils.ts]:
```suggestion
        finally {
            return (simpleJSON && typeof simpleJSON === "object") ? simpleJSON : {}; 
        }
```

File: test/github-issues/4440/issue-4440.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+describe("github issues > #4440 simple-json column type throws error for string with no value", () => {
+
+    let connections: Connection[];
+    before(async () => {
+        connections = await createTestingConnections({
+            entities: [Post],
+            schemaCreate: true,
+            dropSchema: true,
+            enabledDrivers: [
```
Comment (vlapo) [inline test/github-issues/4440/issue-4440.ts]: Removing `enabledDrivers` will enable all drivers. We do not want to update this array every time new driver is added.

File: src/util/DateUtils.ts [hunk 6]
```diff
@@ -170,7 +170,14 @@ export class DateUtils {
...
 
     static stringToSimpleJson(value: any) {
-        return typeof value === "string" ? JSON.parse(value) : value;
+        // try/catch implemented
+        let simpleJSON = null;
+        try {
+            simpleJSON = JSON.parse(value); 
+        } 
+        finally {
+            return (simpleJSON && typeof simpleJSON === "object") ? simpleJSON : {};
```
[the remark this answers is not in the document]
Comment (abhijeet1403) [reply src/util/DateUtils.ts]: makes sense

File: test/github-issues/4440/issue-4440.ts (hunk 5 above)
Comment (abhijeet1403) [reply test/github-issues/4440/issue-4440.ts]: done

File: src/util/DateUtils.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (abhijeet1403) [reply src/util/DateUtils.ts]: Got to know about it since my test was failing

**Comment by abhijeet1403** [comment]
@vlapo @rustamwin Any update on this?

**Comment by abhijeet1403** [comment]
Please merge or close.
Also why is this method written in *dateutils* file in first place?

**Comment by wodka** [comment]
this breaks non object saves

**Pull request merged**